mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -447,6 +474,49 @@ multipartUploadWorker e mpu file a (o, c, i) =
         pure $! Right $! PartResponse i m
 
 
+uploadRecursiveWithMode :: WriteMode -> FilePath -> Address -> EitherT UploadError AWS ()
+uploadRecursiveWithMode m src (Address buck ky) = do
+  es <- tryIO $ getFileStatus src
+  case es of
+    Left _ -> left $ UploadSourceMissing src
+    Right st -> unless (isDirectory st) . left $ UploadSourceNotDirectory src
+  files <- liftIO $ listRecursivelyLocal src
+  let prefixLen = L.length (src </> "a") - 1
+      outputAddrs = fmap (\fp -> Address buck (ky // Key (T.pack $ L.drop prefixLen fp))) files
+  mapM_ (uncurry (uploadWithMode m)) $ L.zip files outputAddrs
+
+-- | Like `listRecursively` but for the local filesystem.
+listRecursivelyLocal :: MonadIO m => FilePath -> m [FilePath]
+listRecursivelyLocal topdir = do
+  entries <- liftIO $ listDirectory topdir
+  (dirs, files) <- liftIO $ partitionM isDir $ fmap (topdir </>) entries
+  others <- concatMapM listRecursivelyLocal dirs
+  pure $ files <> others
+  where
+    isDir :: MonadIO m => FilePath -> m Bool
+    isDir fpath =
+      isDirectory <$> liftIO (getFileStatus fpath)
```

Comment:
I'm not so keen on traversing symlinked directories; it adds a bunch of possible failure modes of the "uploading stuff you didn't mean to" variety. I don't mind supporting traversing symlinks if there's a reason to (e.g., with a flag, like `rsync -L`), but I'd prefer it not be the default behaviour.

Reply:
Yes, agreed.